### ISSUE TITLE
Can ifcopenshell run on the linux arm64?

### ISSUE THREAD

**Issue by Bright-Brave** [IfcOpenShell/IfcOpenShell#3892]
Hi, all
I know the ifcopenshell can works on Win、Mac and Linux, but what about linux arm64?
Anyone can help?

**Comment by aothms** [comment]
You'd have to compile it yourself, which is really quite trivial. All it takes is running `python3 build-all.py` on the host machine. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/nix/build-all.py

**Comment by Bright-Brave** [comment]
Thanks for your quick reply.
Yes, we did encounter lots of issues during compilation using `build-all.py`.
Can you please provide us with a Docker image for the aarch64.

**Comment by aothms** [comment]
We don't have machines to test with, so we won't provide a build. Post the issues on compilation and we'll see how far we get with resolving them.

**Comment by brunopostle** [comment]
IfcOpenShell builds fine on arm64 here, though I have never tested these packages: https://copr.fedorainfracloud.org/coprs/bpostle/IfcOpenShell/build/6449432/

**Comment by Bright-Brave** [comment]
@Moult @brunopostle Thanks for your help. I got it by myself.

**Comment by Moult** [comment]
@Bright-Brave for the benefit of future users, could you help share your experience? Did you run anything differently from the official compilation instructions? Or do the same instructions work without changes? Could you share your build for others to download?

**Comment by sennhvi** [comment]
I have tried rpm package found in this site, but failed to install on my aarch64 system.Now I stuck in the progress of Building executables cause it takes toooo long with all 8G memory consumed, I don't know where it goes and how to handle with it now. Any suggestions are greatly appreciated.

> IfcOpenShell builds fine on arm64 here, though I have never tested these packages: https://copr.fedorainfracloud.org/coprs/bpostle/IfcOpenShell/build/6449432/

**Comment by brunopostle** [comment]
> I have tried rpm package found in this site, but failed to install on my aarch64 system.

These rpm packages will likely only work on fedora Linux, as they are linked with specific versions of opencascade and python (if they are failing to install in fedora then I would like to fix this).

**Comment by sennhvi** [comment]
Thanks for your reply, I tried it on aarch64 but failed, as mentioned in [Note](https://blenderbim.org/docs-python/ifcopenshell/installation.html#installation), seems that I don't need to compile it by myself cause what I need is the ifcopenshell-python part.

**Comment by brunopostle** [comment]
@sennhvi to use the python bindings you will need all parts compiled for aarch64

**Comment by aothms** [comment]
> cause it takes toooo long with all 8G memory consumed

Is it swapping? The build script uses all cores, but in a system with many cores, but little amount of RAM, that could cause issues, because the schema files in ifcparse/ take large amounts of RAM.

You might want to retry with

```bash
export IFCOS_NUM_BUILD_PROCS=1
```

to force it to run on a single thread.

**Comment by sennhvi** [comment]
It's really difficult for me to compile ifcopenshell-python cause I have no experience in this field.

> @sennhvi to use the python bindings you will need all parts compiled for aarch64

**Comment by sennhvi** [comment]
> > cause it takes toooo long with all 8G memory consumed
> 
> Is it swapping? The build script uses all cores, but in a system with many cores, but little amount of RAM, that could cause issues, because the schema files in ifcparse/ take large amounts of RAM.
> 
> You might want to retry with
> 
> ```shell
> export IFCOS_NUM_BUILD_PROCS=1
> ```
> 
> to force it to run on a single thread.

I tried but it not works

**Comment by sennhvi** [comment]
I have compiled it yesterday with errors in log file
[20240108.log](https://github.com/IfcOpenShell/IfcOpenShell/files/13868001/20240108.log)

**Comment by aothms** [comment]
> c++: fatal error: Killed signal terminated program cc1plus
> compilation terminated.

It's still running out of memory.

From the logs you can also see that it is still running multiple compilation threads:

```
[  0%] Building CXX object svgfill/CMakeFiles/libsvgfill.dir/src/svgfill.cpp.o
[  0%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcCharacterDecoder.cpp.o
[  0%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcGlobalId.cpp.o
[  0%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcHierarchyHelper.cpp.o
[  0%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcLogger.cpp.o
[  1%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcParse.cpp.o
[  1%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcSIPrefix.cpp.o
[  1%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcSchema.cpp.o
[  1%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcSpfHeader.cpp.o
[  1%] Building CXX object CMakeFiles/IfcParse.dir/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcUtil.cpp.o
[  1%] Linking CXX static library liblibsvgfill.a
[  1%] Built target libsvgfill
/home/test/Downloads/software/IfcOpenShell/src/ifcparse/IfcParse.cpp: In member function ‘virtual IfcParse::ArgumentList::operator aggregate_of_instance::ptr() const’:
```

You can see that because you get warnings about IfcParse.cpp, but IfcUtil.cpp is the last file for which compilation was started.

build-all.py will show you the `IFCOS_NUM_BUILD_PROCS` used when it starts. Make sure you export that environment variable in the same shell session in which you start the script.

**Comment by sennhvi** [comment]
So glad to receive  your reply.I ran compiling script on remote machine with aarch64 architecture, to make sure the script still run when I not at front  of computer I used command 'screen  -R' and then export the variable IFCOS_NUM_BUILD_PROCS, seems they may have some conflicts here.

Now I'd like to use  x86_64 UBUNTU  to cross compile ,'arm crosschain' may help, not sure whether it works but trying

**Comment by aothms** [comment]
Yes, keep in mind that every screen window is essentially a new shell with their own environment variables.

I don't have a lot of experience with cross compilation myself. The build script uses quite a bit different build systems for the dependencies, eg. boost bjam, autotools, cmake. So it's potentially rather complicated, but we got something similar to work with cross compilation for Web Assembly, so it should be possible.

**Comment by sennhvi** [comment]
I'm not sure if  docker or ARM cross toolchain can help.
